pkg/policy/policies.go

```diff
@@ -58,43 +60,58 @@ func (ps *Policies) PIDFilterMax() uint64 {
 	return ps.pidFilterMax
 }
 
-func (ps *Policies) UIDFilterableInUserSpace() bool {
-	return ps.uidFilterableInUserSpace
+func (ps *Policies) UIDFilterableInUserland() bool {
+	return ps.uidFilterableInUserland
 }
 
-func (ps *Policies) PIDFilterableInUserSpace() bool {
-	return ps.pidFilterableInUserSpace
+func (ps *Policies) PIDFilterableInUserland() bool {
+	return ps.pidFilterableInUserland
 }
 
-// ContainerFilterEnabled returns a bitmask of policies that have at least one container filter type
-// enabled. TODO: make sure the stores are also atomic (an atomic load is only protecting the read
-// from context switches, not from CPU cache coherency issues).
+// ContainerFilterEnabled returns a bitmap of policies that have at least
+// one container filter type enabled.
+//
+// TODO: make sure the stores are also atomic (an atomic load is only protecting
+// the read from context switches, not from CPU cache coherency issues).
 func (ps *Policies) ContainerFilterEnabled() uint64 {
 	return atomic.LoadUint64(&ps.containerFiltersEnabled)
 }
 
-// Compute recalculates values, updates flags and fills the reduced user space
-// map. It must be called at initialization and at every runtime policies changes
+// FilterableInUserland returns a bitmap of policies that must be filtered in userland
+// (ArgFilter, RetFilter, ContextFilter, UIDFilter and PIDFilter).
+//
+// TODO: make sure the stores are also atomic (an atomic load is only protecting
+// the read from context switches, not from CPU cache coherency issues).
```

Comment:
Not that simple (about atomics). If you did something like this:

```
diff --git a/pkg/policy/policies.go b/pkg/policy/policies.go
index 3b363a7d..0f72a4d3 100644
--- a/pkg/policy/policies.go
+++ b/pkg/policy/policies.go
@@ -13,30 +13,49 @@ type Policies struct {
        filterEnabledPoliciesMap  map[*Policy]int      // stores only enabled policies
        filterUserlandPoliciesMap map[*Policy]int      // stores a reduced map with only userland filterable policies
 
-       uidFilterMin            uint64
-       uidFilterMax            uint64
-       pidFilterMin            uint64
-       pidFilterMax            uint64
-       uidFilterableInUserland bool
-       pidFilterableInUserland bool
-
-       filterableInUserland    uint64 // bitmap of policies that must be filtered in userland
-       containerFiltersEnabled uint64 // bitmap of policies that have at least one container filter type enabled
+       uidFilterMin *atomic.Uint64
+       uidFilterMax *atomic.Uint64
+       pidFilterMin *atomic.Uint64
+       pidFilterMax *atomic.Uint64
+
+       uidFilterableInUserland *atomic.Bool
+       pidFilterableInUserland *atomic.Bool
+
+       filterableInUserland    *atomic.Uint64 // bitmap of policies that must be filtered in userland
+       containerFiltersEnabled *atomic.Uint64 // bitmap of policies that have at least one container filter type enabled
 }
 
 func NewPolicies() *Policies {
+       uidFilterMin := &atomic.Uint64{}
+       uidFilterMax := &atomic.Uint64{}
+       pidFilterMin := &atomic.Uint64{}
+       pidFilterMax := &atomic.Uint64{}
+       uidFilterableInUserland := &atomic.Bool{}
+       pidFilterableInUserland := &atomic.Bool{}
+       FilterableInUserland := &atomic.Uint64{}
+       containerFiltersEnabled := &atomic.Uint64{}
+
+       uidFilterMin.Store(filters.MinNotSetUInt)
+       uidFilterMax.Store(filters.MaxNotSetUInt)
+       pidFilterMin.Store(filters.MinNotSetUInt)
+       pidFilterMax.Store(filters.MaxNotSetUInt)
+       uidFilterableInUserland.Store(false)
+       pidFilterableInUserland.Store(false)
+       FilterableInUserland.Store(0)
+       containerFiltersEnabled.Store()
+
        return &Policies{
                policiesArray:             [MaxPolicies]*Policy{},
                filterEnabledPoliciesMap:  map[*Policy]int{},
                filterUserlandPoliciesMap: map[*Policy]int{},
-               uidFilterMin:              filters.MinNotSetUInt,
-               uidFilterMax:              filters.MaxNotSetUInt,
-               pidFilterMin:              filters.MinNotSetUInt,
-               pidFilterMax:              filters.MaxNotSetUInt,
-               uidFilterableInUserland:   false,
-               pidFilterableInUserland:   false,
-               filterableInUserland:      0,
-               containerFiltersEnabled:   0,
+               uidFilterMin:              uidFilterMin,
+               uidFilterMax:              uidFilterMax,
+               pidFilterMin:              pidFilterMin,
+               pidFilterMax:              pidFilterMax,
+               uidFilterableInUserland:   uidFilterableInUserland,
+               pidFilterableInUserland:   pidFilterableInUserland,
+               filterableInUserland:      FilterableInUserland,
+               containerFiltersEnabled:   containerFiltersEnabled,
        }
 }
 
@@ -45,45 +64,39 @@ func (ps *Policies) Count() int {
 }
 
 func (ps *Policies) UIDFilterMin() uint64 {
-       return ps.uidFilterMin
+       return ps.uidFilterMin.Load()
 }
 
 func (ps *Policies) UIDFilterMax() uint64 {
-       return ps.uidFilterMax
+       return ps.uidFilterMax.Load()
 }
 
 func (ps *Policies) PIDFilterMin() uint64 {
-       return ps.pidFilterMin
+       return ps.pidFilterMin.Load()
 }
 
 func (ps *Policies) PIDFilterMax() uint64 {
-       return ps.pidFilterMax
+       return ps.pidFilterMax.Load()
 }
 
 func (ps *Policies) UIDFilterableInUserland() bool {
-       return ps.uidFilterableInUserland
+       return ps.uidFilterableInUserland.Load()
 }
 
 func (ps *Policies) PIDFilterableInUserland() bool {
-       return ps.pidFilterableInUserland
+       return ps.pidFilterableInUserland.Load()
 }
 
-// ContainerFilterEnabled returns a bitmap of policies that have at least
-// one container filter type enabled.
-//
-// TODO: make sure the stores are also atomic (an atomic load is only protecting
-// the read from context switches, not from CPU cache coherency issues).
+// ContainerFilterEnabled returns a bitmap of policies that have at least one container filter type
+// enabled.
 func (ps *Policies) ContainerFilterEnabled() uint64 {
-       return atomic.LoadUint64(&ps.containerFiltersEnabled)
+       return ps.containerFiltersEnabled.Load()
 }
 
-// FilterableInUserland returns a bitmap of policies that must be filtered in userland
-// (ArgFilter, RetFilter, ContextFilter, UIDFilter and PIDFilter).
-//
-// TODO: make sure the stores are also atomic (an atomic load is only protecting
-// the read from context switches, not from CPU cache coherency issues).
+// FilterableInUserland returns a bitmap of policies that must be filtered in userland (ArgFilter,
+// RetFilter, ContextFilter, UIDFilter and PIDFilter).
 func (ps *Policies) FilterableInUserland() uint64 {
-       return atomic.LoadUint64(&ps.filterableInUserland)
+       return ps.filterableInUserland.Load()
 }
 
 // Compute recalculates values, updates flags, fills the reduced userland map,
@@ -98,7 +111,9 @@ func (ps *Policies) Compute() {
        ps.updateContainerFilterEnabled()
 
        userlandMap := make(map[*Policy]int)
-       ps.filterableInUserland = 0
+
+       ps.filterableInUserland.Store(0)
+
        for p := range ps.filterEnabledPoliciesMap {
                if p.ArgFilter.Enabled() ||
                        p.RetFilter.Enabled() ||
```

And encapsulated the Load/Stores for the uint64s, you would still have a problem requiring a mutex protecting those uint64s (and not only atomics). So atomics might not even be needed for them (if they're protected by a mutex).

That happens because you use "utils.SetBit()" in that variable, and the SetBit does:

- load
- shift
- store

Which are 3 operations that can't be protected by a single underlaying HW atomic operation. So you need a mutex (with an atomic primitive under the hood) to protect the load/change/store (as you can't do a "compare and exchange" atomic HW operation for bitwise).

Reply:
Its better if you add a "big lock" (single mutex) and hold that mutex for all operations on your uint64s for now (it will be safer and less error prone). For bools you can use atomics with SetTrue() SetFalse() methods, for uint64s that are simply Load/Store you can also have Set/Get methods, but for uint64s with bits being shifted, you will need a lock if no underlaying atomic instruction/method/operation exists (which I believe is the case here).